joern-cli/frontends/jimple2cpg/src/main/scala/io/joern/jimple2cpg/util/ProgramHandlingUtil.scala

```diff
@@ -1,158 +1,136 @@
 
   private val logger = LoggerFactory.getLogger(ProgramHandlingUtil.getClass)
 
-  /** The temporary directory used to unpack class files to.
-    */
-  private var TEMP_DIR: Option[Path] = None
+  sealed class Entry(entry: Either[File, ZipEntry]) {
 
-  logger.debug(s"Using temporary folder at $TEMP_DIR")
-
-  /** Returns the temporary directory used to unpack and analyze projects in. This allows us to lazily create the
-    * unpacking directory.
-    * @return
-    *   the path pointing to the unpacking directory.
-    */
-  def getUnpackingDir: Path =
-    TEMP_DIR match {
-      case None =>
-        val p = Files.createTempDirectory("joern-")
-        TEMP_DIR = Some(p)
-        p
-      case Some(dir) => dir
-    }
-
-  /** Inspects class files and moves them to the temp directory based on their package path.
-    *
-    * @param files
-    *   the class files to move.
-    * @return
-    *   the list of class files at their new locations.
-    */
-  def moveClassFiles(files: List[String]): List[String] = {
-    var destPath: Option[String] = None
+    def this(file: File) = this(Left(file))
+    def this(entry: ZipEntry) = this(Right(entry))
+    private def file: File            = entry.fold(identity, e => File(e.getName))
+    def name: String                  = file.name
+    def fullExtension: Option[String] = file.extension(includeAll = true)
+    def extension: Option[String]     = file.extension
+    def isDirectory: Boolean          = entry.fold(_.isDirectory, _.isDirectory)
+    def maybeRegularFile(): Boolean   = entry.fold(_.isRegularFile, !_.isDirectory)
+  }
 
-    sealed class ClassPathVisitor extends ClassVisitor(Opcodes.ASM9) {
-      override def visit(
-        version: Int,
-        access: Int,
-        name: String,
-        signature: String,
-        superName: String,
-        interfaces: Array[String]
-      ): Unit = {
-        destPath = Some(getUnpackingDir.toAbsolutePath.toString + File.separator + name + ".class")
-      }
+  private def unfoldArchives[A](src: File, emitOrUnpack: File => Either[A, List[File]]): IterableOnce[A] = {
+    emitOrUnpack(src) match {
+      case Left(a)             => Seq(a)
+      case Right(disposeFiles) => disposeFiles.flatMap(x => unfoldArchives(x, emitOrUnpack))
     }
+  }
+  private def extractClassesToTmp(
+    src: File,
+    tmpDir: File,
+    isArchive: Entry => Boolean,
+    isSource: Entry => Boolean
+  ): IterableOnce[ClassFile] = {
 
-    files.flatMap { f =>
-      Using.resource(new FileInputStream(f)) { fis =>
-        val cr          = new ClassReader(fis)
-        val rootVisitor = new ClassPathVisitor()
-        cr.accept(rootVisitor, SKIP_CODE)
+    def shouldExtract(e: Entry) = e.maybeRegularFile() && (isArchive(e) || isSource(e))
+    unfoldArchives(
+      src,
+      {
+        case f if isSource(Entry(f)) =>
+          Left(ClassFile(f))
+        case f if f.isDirectory() =>
+          val files = f.listRecursively.filter(!_.isDirectory).toList
```

Comment:
```suggestion
          val files = f.listRecursively.filterNot(_.isDirectory).toList
```

Reply:
I hate that this function exists, but will use it if it is available.